pkg/encode/encoder.go

```diff
@@ -0,0 +1,451 @@
+	case []byte:
+		return e.writeBytes(v)
+	case []interface{}:
+		return e.writeList(v)
+	case map[string]interface{}:
+		return e.writeMap(v)
+	}
+
+	// 构造并编码Java对象
+	return e.writeClass(clazz)
+}
+
+// 获取Java对象的类型信息
+func getClass(obj Object) (*Class, error) {
+	if obj == nil {
+		return nil, fmt.Errorf("getClass: nil object")
+	}
+
+	// 获取对象的类型
+	t := reflect.TypeOf(obj)
+
+	// 获取对象的类名
+	name := getClassName(t)
+	if name == "" {
+		return nil, fmt.Errorf("getClass: invalid object type: %v", t)
+	}
+
+	// 如果是struct类型，则根据结构体字段构造Class对象
+	if t.Kind() == reflect.Struct {
+		fields, err := getClassFields(t)
+		if err != nil {
+			return nil, err
+		}
+		return &Class{Name: name, Fields: fields}, nil
+	}
+
+	// 其他基本类型直接构造Class对象
+	return &Class{Name: name}, nil
+}
+
+// 获取对象的类名
+func getClassName(t reflect.Type) string {
+	switch t.Kind() {
+	case reflect.Struct:
+		if t.NumField() == 0 {
+			return ""
+		}
+		// 结构体类型的类名为结构体名的小写形式
+		return strings.ToLower(t.Name())
+	case reflect.Slice:
+		// 判断是否为字节数组
+		if t.Elem().Kind() == reflect.Uint8 {
+			return "byte[]"
+		}
+		// 其他slice类型的类名为List
+		return "list"
+	case reflect.Map:
+		// map类型的类名为Map
+		return "map"
+	case reflect.Ptr:
+		// 指针类型递归调用获取类名
+		return getClassName(t.Elem())
+	default:
+		// 其他基本类型的类名为类型名的小写形式
+		return strings.ToLower(t.Name())
+	}
+}
+
+// 将Java对象类型信息编码为commons协议格式
+func (e *Encoder) writeClass(clazz *Class) error {
+	// 查找类型引用
+	if ref, ok := e.findTypeRef(clazz); ok {
+		return e.writeTypeRef(ref)
+	}
+
+	// 将类型添加到引用数组
+	e.typeRefs = append(e.typeRefs, clazz)
+
+	// 写入类型标识
+	if err := e.WriteByte(byte('C')); err != nil {
+		return err
+	}
+
+	// 写入类名
+	if err := e.WriteString(clazz.Name); err != nil {
+		return err
+	}
+
+	// 写入类型信息
+	if err := e.writeFields(clazz.Fields); err != nil {
+		return err
+	}
+
+	// 标记已经写入类型信息
+	e.typeRefWritten = true
+
+	return nil
+}
+
+// 查找类型引用
+func (e *Encoder) findTypeRef(clazz *Class) (int64, bool) {
+	for i, ref := range e.typeRefs {
+		if reflect.DeepEqual(clazz, ref) {
+			return int64(i), true
+		}
+	}
+	return -1, false
+}
+
+// 写入类型引用
+func (e *Encoder) writeTypeRef(ref int64) error {
+	if err := e.WriteByte(byte('T')); err != nil {
+		return err
+	}
+	if err := e.WriteInt(ref); err != nil {
+		return err
+	}
+	return nil
+}
+
+// 写入字段信息
+func (e *Encoder) writeFields(fields []*Field) error {
+	for _, field := range fields {
+		if err := e.writeField(field); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// 写入单个字段信息
+func (e *Encoder) writeField(field *Field) error {
+	if err := e.WriteString(field.Name); err != nil {
+		return err
+	}
+	if err := e.writeTypeRefOrClass(field.Type); err != nil {
+		return err
+	}
+	if field.Default != nil {
+		if err := e.WriteObject(field.Default); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// 写入类型引用或类名
+func (e *Encoder) writeTypeRefOrClass(clazz *Class) error {
+	if ref, ok := e.findTypeRef(clazz); ok {
+		if err := e.writeTypeRef(ref); err != nil {
```

Comment:
it will be simpler to just `return e.writeTypeRef(ref)`, also  for the else branch.